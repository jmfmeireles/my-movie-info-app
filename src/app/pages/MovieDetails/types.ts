export interface MoviesDetailsState {
  isCallingAPI: boolean;
  movieDetailedData?: any;
}

export interface MovieDetailedData {
  id: string;
  title: string;
  originalTitle: string;
  fullTitle: string;
  year: string;
  image: string;
  releaseDate: string;
  runtimeStr: string;
  plot: string;
  plotLocal: string;
  awards: string;
  directorList: CastMemberInfo[];
  writerList: CastMemberInfo[];
  directors: string;
  writers: string;
  starList: CastMemberInfo[];
  actorList: Actor[];
  genreList: KeyValuePair[];
  imDbRating: string;
  imDbRatingVotes: string;
  trailer: Trailer;
}

export interface CastMemberInfo {
  id: string;
  name: string;
}

export interface Actor {
  id: string;
  image: string;
  name: string;
  asCharacter: string;
}

export interface Trailer {
  thumbnailUrl: string | null;
  link: string | null;
  linkEmbed: string | null;
  errorMessage: string;
}

export interface GetMovieDetailsPayload {
  movieNumber: string;
  languageCode: string;
}

export interface KeyValuePair {
  key: string;
  value: string;
}
