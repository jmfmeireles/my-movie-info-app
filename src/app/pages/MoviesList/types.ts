export interface MoviesListState {
  isCallingAPI: boolean;
  movies: MovieMetadata[];
  pagination: Pagination;
  filter: Filter;
  sorter: Sorter;
}

export interface MovieMetadata {
  id: string;
  rank: string;
  title: string;
  fullTitle: string;
  year: string;
  image: string;
  crew: string;
  imDbRating: string;
  imDbRatingCount: string;
}

export enum NumberKeys {
  RANK = 'rank',
  IMDB_RATING = 'imDbRating',
  YEAR = 'year',
  IMDB_RATING_COUNT = 'imDbRatingCount',
}

export interface Pagination {
  page: number;
  elementsPerPage: number;
  numberOfPossibleResults: number;
}

export interface Filter {
  title: string;
}

export interface Sorter {
  sortBy: SorterCategory;
  direction: SorterDirection;
}

export enum SorterCategory {
  RANK = 'rank',
  TITLE = 'title',
  YEAR = 'year',
  IMDB_RATING_COUNT = 'imDbRatingCount',
}

export enum SorterDirection {
  ASCENDING = 'ascending',
  DESCENDING = 'descending',
}
