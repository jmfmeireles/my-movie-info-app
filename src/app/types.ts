export interface PersonData {
  name: string;
  surname: string;
  birthdate: Date | null;
  country: string;
}

export interface FormDataState {
  personRecords: PersonData[];
}
