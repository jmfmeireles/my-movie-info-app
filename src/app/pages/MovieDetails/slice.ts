import { PayloadAction } from 'typesafe-actions';
import { createSlice } from 'utils/@reduxjs/toolkit';
import {
  MovieDetailedData,
  MoviesDetailsState,
  GetMovieDetailsPayload,
} from './types';

export const initialState: MoviesDetailsState = {
  isCallingAPI: false,
};

const movieDetailsSlice = createSlice({
  name: 'movieDetails',
  initialState,
  reducers: {
    /**
     * Start action of this page, which triggers a saga to get the movie details from API
     */
    initMovieDetailsPage(
      state: MoviesDetailsState,
      action: PayloadAction<string, GetMovieDetailsPayload>,
    ) {
      state.isCallingAPI = true;
    },
    /**
     * Action triggered after a successfull call to get movie details from API
     */
    setMovieDetails(
      state: MoviesDetailsState,
      action: PayloadAction<string, MovieDetailedData>,
    ) {
      state.isCallingAPI = false;
      state.movieDetailedData = action.payload;
    },
  },
});

export const { actions, reducer, name: sliceKey } = movieDetailsSlice;
