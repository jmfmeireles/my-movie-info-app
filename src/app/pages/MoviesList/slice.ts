import { PayloadAction } from 'typesafe-actions';
import { createSlice } from 'utils/@reduxjs/toolkit';
import Movies from './resources/helpers/Movies';
import {
  Filter,
  MovieMetadata,
  MoviesListState,
  Sorter,
  SorterCategory,
  SorterDirection,
} from './types';

export const initialState: MoviesListState = {
  isCallingAPI: false,
  movies: [],
  pagination: {
    page: 0,
    elementsPerPage: 12,
    numberOfPossibleResults: 0,
  },
  filter: {
    title: '',
  },
  sorter: {
    sortBy: SorterCategory.RANK,
    direction: SorterDirection.ASCENDING,
  },
};

const moviesListSlice = createSlice({
  name: 'moviesList',
  initialState,
  reducers: {
    /**
     * Start action of this page, which triggers a saga to get the products from API
     */
    initMoviesListPage(
      state: MoviesListState,
      action: PayloadAction<string, string>,
    ) {
      state.pagination = initialState.pagination;
      state.isCallingAPI = true;
    },
    /**
     * Action triggered after a successfull call to get movies from API
     */
    setMovies(
      state: MoviesListState,
      action: PayloadAction<string, MovieMetadata[]>,
    ) {
      state.isCallingAPI = false;
      state.movies = action.payload;
      state.pagination.numberOfPossibleResults = action.payload.length;
    },
    /****************************PAGINATION ****************************************/
    setPage(state: MoviesListState, action: PayloadAction<string, number>) {
      state.pagination.page = action.payload;
    },
    /**************************** FILTERS AND SORTERS ****************************************/
    setSorterData(
      state: MoviesListState,
      action: PayloadAction<string, Sorter>,
    ) {
      state.sorter = action.payload;
    },
    setFilterData(
      state: MoviesListState,
      action: PayloadAction<string, Filter>,
    ) {
      state.filter = action.payload;
      state.pagination.numberOfPossibleResults =
        Movies.getNumberOfProductsAfterFiltering(state.movies, action.payload);
    },
  },
});

export const { actions, reducer, name: sliceKey } = moviesListSlice;
