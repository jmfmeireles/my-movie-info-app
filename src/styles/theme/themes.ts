import { createTheme } from '@mui/material/styles';

const lightTheme = {
  primary: '#EDBF17',
  textPrimary: '#666666',
  textSecondary: 'rgba(58,52,51,0.7)',
  header: '#ffffff',
  footer: '#ffffff',
  background: '#F8F8F8',
  backgroundVariant: 'rgba(251,249,249,1)',
  border: 'rgba(58,52,51,0.12)',
  borderLight: 'rgba(58,52,51,0.05)',
};

export type Theme = typeof lightTheme;

export const themes = {
  light: lightTheme,
};

export const materialUiTheme = createTheme({
  typography: {
    fontFamily: ['"Gruppo"', 'sans-serif'].join(','),
  },
  components: {
    MuiChip: {
      styleOverrides: {
        avatar: {
          backgroundColor: lightTheme.primary,
          color: '#fff',
        },
      },
    },
  },
  palette: {
    primary: { main: lightTheme.primary },
  },
});
