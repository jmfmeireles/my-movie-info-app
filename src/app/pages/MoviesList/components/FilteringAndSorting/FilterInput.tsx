import React, { ChangeEvent, useState } from 'react';

import { useSelector, useDispatch } from 'react-redux';
import { useTranslation } from 'react-i18next';
import InputBase from '@mui/material/InputBase';
import IconButton from '@mui/material/IconButton';
import SearchIcon from '@mui/icons-material/Search';

import { Filter } from '../../types';
import { getFilter } from '../../selectors';
import { actions as movieListActions } from '../../slice';
import { FilterWrapper } from './styles';

export default function FilterInput() {
  const { t } = useTranslation();
  const dispatch = useDispatch();

  const filterData: Filter = useSelector(getFilter);

  const [titleFilter, setTitleFilter] = useState<string>(filterData.title);

  const handleFilterChange = (event: ChangeEvent<HTMLInputElement>) => {
    const stringToFilter: string = event.target.value;
    setTitleFilter(stringToFilter);
    if (!stringToFilter) {
      dispatch(
        movieListActions.setFilterData({
          title: '',
        }),
      );
    }
  };

  const handleKeyDown = event => {
    if (event.code === 'Enter') {
      handleSearch();
    }
  };

  const handleSearch = () => {
    dispatch(
      movieListActions.setFilterData({
        title: titleFilter,
      }),
    );
  };
  return (
    <FilterWrapper>
      <IconButton onClick={handleSearch} data-testid="searchIcon">
        <SearchIcon />
      </IconButton>
      <InputBase
        inputProps={{ 'data-testid': 'searchByTitle' }}
        placeholder={t('moviesListPage.searchByTitle')}
        onChange={handleFilterChange}
        onKeyDown={handleKeyDown}
      />
    </FilterWrapper>
  );
}
