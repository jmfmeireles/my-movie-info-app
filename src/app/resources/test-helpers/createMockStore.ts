import { RootState } from '../../../types';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';
import lodash, { cloneDeep } from 'lodash';
import { initialState as themeState } from '../../../styles/theme/slice';
import { initialState as moviesListState } from '../../pages/MoviesList/slice';
import { initialState as movieDetailsState } from '../../pages/MovieDetails/slice';

export default function createMockStore(
  modifyState?: (state: RootState) => RootState,
): MockStoreEnhanced<RootState, {}> {
  const initialState = cloneDeep<RootState>({
    theme: themeState,
    moviesList: moviesListState,
    movieDetails: movieDetailsState,
  });
  const mockStore = configureStore<RootState>([]);
  const updatedState = modifyState ? modifyState(initialState) : initialState;
  return mockStore(lodash.cloneDeep(updatedState));
}
