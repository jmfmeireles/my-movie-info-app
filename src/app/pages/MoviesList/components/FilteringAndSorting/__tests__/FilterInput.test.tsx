import React from 'react';

import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { fireEvent, render } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';

import FilterInput from '../FilterInput';
import createMockStore from '../../../../../resources/test-helpers/createMockStore';
import { RootState } from '../../../../../../types';
import { themes } from '../../../../../../styles/theme/themes';
import { actions as movieListActions } from '../../../slice';

describe('List', () => {
  configure({ adapter: new Adapter() });

  const store: MockStoreEnhanced<RootState, {}> = createMockStore(state => {
    if (state.moviesList) {
      state.moviesList = {
        ...state.moviesList,
        filter: {
          title: 'title',
        },
      };
    }
    return state;
  });

  it('Renders properly', () => {
    expect(
      mount(
        <ThemeProvider theme={themes.light}>
          <BrowserRouter>
            <Provider store={store}>
              <FilterInput />
            </Provider>
          </BrowserRouter>
        </ThemeProvider>,
      ),
    ).toMatchSnapshot();
  });
  it('Searchs a value properly', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={themes.light}>
        <BrowserRouter>
          <Provider store={store}>
            <FilterInput />
          </Provider>
        </BrowserRouter>
      </ThemeProvider>,
    );

    const searchInput = getByTestId(`searchByTitle`);
    fireEvent.change(searchInput, { target: { value: 'test' } });
    const searchIcon = getByTestId(`searchIcon`);
    fireEvent.click(searchIcon);
    expect(store.getActions()).toMatchObject([
      {
        type: movieListActions.setFilterData.type,
        payload: {
          title: 'test',
        },
      },
    ]);
  });
});
