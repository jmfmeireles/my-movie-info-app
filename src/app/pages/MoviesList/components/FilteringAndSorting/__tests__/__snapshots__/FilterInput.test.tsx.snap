// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Renders properly 1`] = `ReactWrapper {}`;
